### DISCUSSION TITLE
Calculating percent error expressed in geoduck transcriptome

### DISCUSSION THREAD

**Discussion by ocattau** [RobertsLab/resources#1398] [Q&A]
Lets say my data looks like this table after the count data has been aligned with the blast data. In order to calculate % error expression, my first reaction would be to calculate % NA in the Identifier column, but sometimes there is count data without an identifier. 

   My second guess would be to filter the tpm so only values > 0 are counted, then calculate the % that has any value at all, but would some values be NA and some 0? 

   I don't think it makes any sense to calculate % error expressed using the Target_ID column since most of them have values but sometimes there are identifiers that don't have a corresponding Target_ID.  

   If I use the R function complete.cases, it removed any row with NAs which would delete all rows 1-3 in this example, and in my mind if there is count data but no identifiers, wouldn't that mean the gene exists but we haven't identified it yet? Any help with this data wrangling would be appreciated. 

| Target_ID | Identifiers | tpm |
|---|---|---|
|NA|BJC123|NA|
|Trinity-X12| NA | 20|
|Trinity-Y99| NA | 0 |
|Trinity-Z33| ABC321 | 3000 |
|Trinity-Z34| ABC321 | 2000 |

**Comment by sr320** [comment]
Can you explain what you mean by "% error expression"?

**Reply by ocattau** [reply]
I was thinking about it as  "The accuracy (of the quantification) relies on there being enough unique information in the reads to enable bioinformatics tools to accurately assign the reads to the correct gene." So potentially that could be the percentage of target_IDs from the countMatrix that have identifiers and counts vs NAs?

**Reply by kubu4** [reply]
> The accuracy (of the quantification) relies on there being enough unique information in the reads to enable bioinformatics tools to accurately assign the reads to the correct gene.

This is precisely what assembly/alignment algorithms are designed to handle. So, the transcriptome assembly that you're using has already utilized an algorithm to determine which reads can be aligned to each other and whether or not there is enough evidence (e.g. # of reads with same sequence) to align reads to generate a longer contig (i.e. gene) from the consensus of the aligned reads at each nucleotide position.

For most (all?) aligners, you can change various settings to adjust the alignment stringency. For example, you could probably set the stringency to a very strict level where only reads which had 100% matches at each nucleotide position would be aligned/counted. Then, you'd have 0% "error" in expression (although, I'm still not totally clear on what you mean by this). A stringency that strict would not allow for any biological variation (e.g. SNPs) in your samples and so you'd actually be severely under-counting expression in this scenario.

Bigger question is why are you interested in calculating a metric like this? What are you hoping to learn from it?

**Reply by ocattau** [reply]
Besides Steven asked me to calculate it? The goal is to characterize the transcriptome data in a quantitative way. Maybe calculating the % of Target_IDs matched to identifiers off Swiss Prot seems relevant to me for this question?

**Reply by sr320** [reply]
I asked to calculate the percentage of genes expressed in each tissue... among other things. Not anything to do with error..

**Reply by ocattau** [reply]
So I should just ignore any TargetID without an identifier? (for example I would delete the first three rows in this sample data set)

**Reply by sr320** [reply]
You keep every Target ID in the table I provided

**Reply by ocattau** [reply]
In that case 95% of the data set is NA, is that normal?

**Reply by sr320** [reply]
You want to know the expression pattern of every Target ID, including those that are not expressed at all.

**Reply by sr320** [reply]
so you really have a line like line 1? no target ID?

**Comment by sr320** [comment]
I think I see the problem 

You should be using left_join (or something similar) where you retain only annotation information that matches data table with expression

**Reply by ocattau** [reply]
Yes but I think that is just because I downloaded all reviewed proteins from uniprot for go terms

**Reply by ocattau** [reply]
Left_join is the same thing as complete.cases() and removes row two and three. I thought we cared about all count data regardless if it has a Goterm associated with it.

**Reply by ocattau** [reply]
![Screen Shot 2022-02-10 at 1 46 39 PM](https://user-images.githubusercontent.com/81712104/153501908-c1c623e0-7211-4311-a8a3-272e7a43f26e.png)

**Reply by sr320** [reply]
we do - but your line 1 has no count data.

**Reply by sr320** [reply]
left join would not remove line 2 and 3.

**Reply by ocattau** [reply]
https://github.com/ocattau/code/blob/main/characterize_larval_transciptome/inspecting%20kallisto%20output%20SR%20merged%20kallisto.Rmd

**Reply by ocattau** [reply]
Okay I figured it out, I did need to add a left.join at line 42 of the code I posted above, thanks!

**Comment by sr320** [comment]
Put a link to the GitHub repo where you are working.. that would be easiest way to understand and troubleshoot

**Reply by sr320** [reply]
I will play with some on repo, but I do think in one of your left joins the count matrix needs to be first. The final table should the same number of lines as count matrix.

**Reply by sr320** [reply]
This code looks good to me... https://d.pr/n/f0geiK

Looking at this --- the fact there are 1M transcripts and only 70k blast hits is of a first concern.. Where is this code?

**Reply by sr320** [reply]
@ocattau Can you provide link to blast code?

**Reply by ocattau** [reply]
I agree that looks weird, https://github.com/ocattau/ocattau.github.io/blob/master/_posts/2021-10-21-functiongeneannotations.md

**Reply by sr320** [reply]
if this is it
```

/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/blastx \
-query /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-v1.0.fa \
-db /gscratch/srlab/ocattau/p.generosa/Pgenerosa_transcriptome_v5.fasta
-out /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 30 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6
```

needs revisiting

**Reply by ocattau** [reply]
What's wrong with it? Did it run out of time do you think?

**Reply by sr320** [reply]
code I posted from your repo is genome v trans. (you really should organize repo).

looking at link you provided 

change evalue

to `evalue 1E-05`

and make sure it finishes.

**Reply by ocattau** [reply]
Okay I'll redo the blast, remember when we did this? https://github.com/RobertsLab/resources/discussions/1324

**Reply by ocattau** [reply]
What do you make of this? "QOSGrpMemLimit"
![Screen Shot 2022-02-10 at 4 51 33 PM](https://user-images.githubusercontent.com/81712104/153521544-66f0a845-8cad-429f-a86f-d4108e503feb.png)

**Reply by sr320** [reply]
what does you slurm header look like?

**Reply by ocattau** [reply]
![Screen Shot 2022-02-10 at 4 58 17 PM](https://user-images.githubusercontent.com/81712104/153522014-43624072-4f4a-4f00-980c-938631869a5e.png)

**Reply by ocattau** [reply]
I think I need to wait till Sam's job is done because of memory issues

**Reply by sr320** [reply]
change mem to 100G

**Reply by sr320** [reply]
Full explanation - we have two nodes
https://robertslab.github.io/resources/Computing-Hardware/#mox-hyak

Sam is using the 500M, but you do not require that. When you set mem  to <120G your job will start on 2nd node.
